tensorflow/core/kernels/matmul_op_impl.h

```diff
@@ -357,31 +531,164 @@ struct LaunchBatchMatMul<GPUDevice, Scalar> {
         std::min(bcast.x_batch_size(), bcast.y_batch_size()) == 1;
-    bool use_strided_batched =
-        (!bcast.IsBroadcastingRequired() || is_full_broadcast) &&
-        batch_size > 1;
-    if (use_strided_batched) {
-      a_stride = bcast.x_batch_size() != 1 ? m * k : 0;
-      b_stride = bcast.y_batch_size() != 1 ? k * n : 0;
+    bool requires_mixed_broadcasting =
+        bcast.IsBroadcastingRequired() && !is_full_broadcast;
+    if (!requires_mixed_broadcasting) {
+      bool broadcast_a = bcast.x_batch_size() == 1;
+      bool broadcast_b = bcast.y_batch_size() == 1;
+      a_stride = broadcast_a ? 0 : m * k;
+      b_stride = broadcast_b ? 0 : k * n;
       c_stride = m * n;
       a_device_memory.push_back(AsDeviceMemory(a_base_ptr));
       b_device_memory.push_back(AsDeviceMemory(b_base_ptr));
       c_device_memory.push_back(AsDeviceMemory(c_base_ptr));
       a_ptrs.push_back(&a_device_memory.back());
       b_ptrs.push_back(&b_device_memory.back());
       c_ptrs.push_back(&c_device_memory.back());
-    } else if (!bcast.IsBroadcastingRequired()) {
-      for (int64 i = 0; i < batch_size; ++i) {
-        a_device_memory.push_back(AsDeviceMemory(a_base_ptr + i * m * k));
-        b_device_memory.push_back(AsDeviceMemory(b_base_ptr + i * k * n));
-        c_device_memory.push_back(AsDeviceMemory(c_base_ptr + i * m * n));
-        a_ptrs.push_back(&a_device_memory.back());
-        b_ptrs.push_back(&b_device_memory.back());
-        c_ptrs.push_back(&c_device_memory.back());
+
+      DataType dtype = DataTypeToEnum<Scalar>::value;
+      bool allow_tf32 = tensor_float_32_execution_enabled();
+      int device_id = stream->parent()->device_ordinal();
+      BatchMatmulParameters matmul_parameters(
+          trans_x, trans_y, adj_x, adj_y, m, n, k, batch_size, broadcast_a,
+          broadcast_b, dtype, dtype, allow_tf32, device_id);
+
+      static const bool max_autotune_algorithm_count =
+          MatmulMaxAutotuneAlgorithmCount();
+      int max_algorithm_count = use_autotune ? max_autotune_algorithm_count : 1;
+
+      const auto* plan_and_algorithms =
+          BatchMatmulPlanMapSingleton::GetInstance()->Find(matmul_parameters);
+      if (!plan_and_algorithms) {
+        se::blas::DataType blas_dtype = se::blas::ToDataType<Scalar>::value;
+        se::blas::ComputationType computation_type;
+        OP_REQUIRES(
+            context,
+            GetBlasComputationType(dtype, allow_tf32, &computation_type),
+            errors::Internal("Unsupported dtype for batched matmul"));
+
+        se::blas::BlasLtMatmulPlanParams plan_params;
+        plan_params.ab_type = blas_dtype;
+        plan_params.c_type = blas_dtype;
+        plan_params.computation_type = computation_type;
+        plan_params.pointer_mode = se::blas::PointerMode::kHost;
+        plan_params.epilogue = se::blas::Epilogue::kDefault;
+        plan_params.transa = blas_transpose_b;
+        plan_params.transb = blas_transpose_a;
+        plan_params.m = n;
```

Comment:
Why is `transa`, transb`, `m` and `n` flipped here? Essentially it seems they are flipped when creating matmul_paramters. I don't understand why that needs to be that way. `m` is essentially the num_rows and `n` the num columns of the output.

Reply:
XLA doesn't see any conjugate transpose. IIUC can we just use `plan_params.m = m;` etc. where `int64 m = output_matrix.num_cols;` in XLA thunk.

Reply:
Also, it seems `plan_params.lda = lhs_matrix.num_rows;` ,`plan_params.ldb = rhs_matrix.num_rows;` and `plan_params.ldc = output_matrix.num_rows;`

Reply:
> @benbarsdell Any update on this PR? Please. Thanks!

@gbaned @benbarsdell has clarified my doubts offline. Thanks for following up though.